### ISSUE TITLE
Does this achieve benchmarks in paper?

### ISSUE THREAD

**Issue by akshilpatel** [vtalpaert/pytorch-feudal-network#1]
Hi,

Given that the original repo by dmakian doesn't achieve the benchmarks on ATARI etc. given in the papers, I was wondering if this implementation does?

Thanks,

Akshil

**Comment by vtalpaert** [comment]
Hi Akshil,

I just wrote this as an exercise to get to know pytorch. The code is old compared to the latest pytorch.

I'm sure it can get results with minimum tweeking. For example, in https://github.com/vtalpaert/pytorch-feudal-network/blob/master/fun.py#L43 the class

```python
class dLSTM(nn.Module):
    """Implements the dilated LSTM
    Uses a cyclic list of size r to keep r independent hidden states
    """
    def __init__(self, r, input_size, hidden_size):
        super(dLSTM, self).__init__()
        self.lstm = nn.LSTMCell(input_size, hidden_size)
        self.r = r

    def init_state(self, batch_size):
        # note that we cannot keep the state in only one tensor as updating one place of the tensor counts
        # as an inplace operation and breaks the gradient history
        h0 = [torch.zeros(batch_size, self.lstm.hidden_size, requires_grad=self.training) for _ in range(self.r)]
        c0 = [torch.zeros(batch_size, self.lstm.hidden_size, requires_grad=self.training) for _ in range(self.r)]
        tick = 0
        return tick, h0, c0

    def forward(self, inputs, states):
        """Returns g_t, (tick, hidden)
        """
        tick, hx, cx = states
        hx[tick], cx[tick] = self.lstm(inputs, (hx[tick], cx[tick]))
        tick = (tick + 1) % self.r
        out = sum(hx) / self.r  # TODO verify that network output is mean of hidden states
        return out, (tick, hx, cx)
```

should use a `nn.LSTMCell(r * input_size, r * hidden_size)` and then in forward create a mask or use [torch.masked_select](https://pytorch.org/docs/stable/torch.html#torch.masked_select).

I believe what doesn't work in other pytorch implementations and mine's is the use of a list that breaks the gradient somehow.
